### PR TITLE
support PeftMixedModel signature inspect

### PR DESCRIPTION
**Pull request by Facico**
support PeftMixedModel signature inspect

Use model.base_model.model to get the base model(PeftMixedModel don't have "get_base_model" attribute)

### CODE REVIEW

**Comment by ArthurZucker** [comment]
cc @younesbelkada

**Review by younesbelkada — COMMENTED** [review]
Thanks for adding the support for `PeftMixedModel`! Since that feature is only available from peft >= 0.7.0, could you import the `PeftMixedModel` only if PEFT version is greater or equal than 0.7.0? https://github.com/huggingface/peft/releases/tag/v0.7.0 
You can use `version.parse` e.g.: https://github.com/huggingface/transformers/blob/3eddda1111f70f3a59485e08540e8262b927e867/src/transformers/modeling_utils.py#L2807

**Comment by Facico** [comment]
@younesbelkada Thanks for the suggestion, I've pushed the changes!

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -185,9 +185,12 @@
 if is_safetensors_available():
     import safetensors.torch
 
-
+is_peftmix_available = False
```
Comment (younesbelkada) [inline src/transformers/trainer.py:188]:
```suggestion
```

File: src/transformers/trainer.py [hunk 2]
```diff
@@ -185,9 +185,12 @@
...
 if is_peft_available():
     from peft import PeftModel
+    is_peftmix_available = version.parse(importlib.metadata.version("peft")) >= version.parse("0.7.0")
+    if is_peftmix_available:
+        from peft import PeftMixedModel
```
Comment (younesbelkada) [inline src/transformers/trainer.py:193]:
```suggestion
```

File: src/transformers/trainer.py [hunk 3]
```diff
@@ -213,7 +216,7 @@
 
 
 def _is_peft_model(model):
-    return is_peft_available() and isinstance(model, PeftModel)
+    return is_peft_available() and isinstance(model, PeftModel) or is_peft_available() and is_peftmix_available and isinstance(model, PeftMixedModel)
```
Comment (younesbelkada) [inline src/transformers/trainer.py:219]:
```suggestion
    classes_to_check = (PeftModel,)
    if version.parse(importlib.metadata.version("peft")) >= version.parse("0.7.0"):
        from peft import PeftMixedModel
        classes_to_check = (*classes_to_check, PeftMixedModel)
    return is_peft_available() and isinstance(model, classes_to_check)
```

**Review by younesbelkada — COMMENTED** [review]
Thanks for iterating, I left an alternative suggestion, could you let me know if this works ?

File: src/transformers/trainer.py [hunk 4]
```diff
@@ -213,7 +212,11 @@
 
 
 def _is_peft_model(model):
-    return is_peft_available() and isinstance(model, PeftModel)
+    classes_to_check = (PeftModel,)
+    if version.parse(importlib.metadata.version("peft")) >= version.parse("0.7.0"):
```
Comment (younesbelkada) [inline src/transformers/trainer.py:216]:
```suggestion
    # Here we also check if the model is an instance of `PeftMixedModel` introduced in peft>=0.7.0: https://github.com/huggingface/transformers/pull/28321
    if version.parse(importlib.metadata.version("peft")) >= version.parse("0.7.0"):
```

**Review by younesbelkada — COMMENTED** [review]
Left one last comment!

**Comment by younesbelkada** [comment]
@Facico can you confirm the latest commit on this branch fixes your issue?

**Comment by Facico** [comment]
Yes, They all work. @younesbelkada

File: src/transformers/trainer.py [hunk 5]
```diff
@@ -213,7 +212,12 @@
 
 
 def _is_peft_model(model):
-    return is_peft_available() and isinstance(model, PeftModel)
+    classes_to_check = (PeftModel,)
```
Comment (younesbelkada) [inline src/transformers/trainer.py:215]:
```suggestion
    classes_to_check = (PeftModel,) if is_peft_available() else ()
```

File: src/transformers/trainer.py [hunk 6]
```diff
@@ -213,7 +212,12 @@
...
+    # Here we also check if the model is an instance of `PeftMixedModel` introduced in peft>=0.7.0: https://github.com/huggingface/transformers/pull/28321
+    if version.parse(importlib.metadata.version("peft")) >= version.parse("0.7.0"):
```
Comment (younesbelkada) [inline src/transformers/trainer.py:217]:
```suggestion
    if is_peft_available() and version.parse(importlib.metadata.version("peft")) >= version.parse("0.7.0"):
```

**Review by younesbelkada — COMMENTED** [review]
Thanks for iterating! The suggestions below shiuld solve the failing CI, can you also run the styling checks? `make fixup`

**Review by younesbelkada — COMMENTED** [review]
Thanks for iterating @Facico ! Can you tun the styling checks? just run `make fixup` and push the formatted file

File: src/transformers/trainer.py [hunk 7]
```diff
@@ -700,7 +705,11 @@ def _set_signature_columns_if_needed(self):
             # Inspect model forward signature to keep only the arguments it accepts.
             model_to_inspect = self.model
             if _is_peft_model(self.model):
-                model_to_inspect = self.model.get_base_model()
+                if hasattr(self.model, "get_base_model"):
+                    model_to_inspect = self.model.get_base_model()
+                else:
+                    # PeftMixedModel
```
Comment (younesbelkada) [inline src/transformers/trainer.py:711]:
```suggestion
                    # PeftMixedModel do not have `get_base_model` method
```

**Review by younesbelkada — APPROVED** [review]
Thanks for adding `PeftMixedModel` support in HF trainer! failing tests are unrelated to the PR

File: src/transformers/trainer.py [hunk 8]
```diff
@@ -213,7 +212,13 @@
...
 
 def _is_peft_model(model):
-    return is_peft_available() and isinstance(model, PeftModel)
+    classes_to_check = (PeftModel,) if is_peft_available() else ()
+    # Here we also check if the model is an instance of `PeftMixedModel` introduced in peft>=0.7.0: https://github.com/huggingface/transformers/pull/28321
+    if is_peft_available() and version.parse(importlib.metadata.version("peft")) >= version.parse("0.7.0"):
+        from peft import PeftMixedModel
+
+        classes_to_check = (*classes_to_check, PeftMixedModel)
+    return is_peft_available() and isinstance(model, classes_to_check)
```
Comment (ArthurZucker) [inline src/transformers/trainer.py:221]:
```suggestion
    if is_peft_available():
        classes_to_check = (PeftModel,) if is_peft_available() else ()
        # Here we also check if the model is an instance of `PeftMixedModel` introduced in peft>=0.7.0: https://github.com/huggingface/transformers/pull/28321
        if version.parse(importlib.metadata.version("peft")) >= version.parse("0.7.0"):
            from peft import PeftMixedModel
            classes_to_check = (*classes_to_check, PeftMixedModel)
         return isinstance(model, classes_to_check)
    return False
```
let's check for peft lib only once 😉

**Review by ArthurZucker — APPROVED** [review]
Thanks, left a small nit

**Comment by Facico** [comment]
@younesbelkada How do I pass the tests_exotic_models?

**Comment by ArthurZucker** [comment]
Rebasing on main should help!

**Comment by younesbelkada** [comment]
Yes @Facico please merge your branch with upstream main and the test should be fixed. Afrer that we'll be able to merge your PR

**Comment by Facico** [comment]
@younesbelkada @ArthurZucker Thanks! That helped.

**Review by younesbelkada — APPROVED** [review]
Thanks again !

**Pull request merged**